core/src/main/java/org/jclouds/http/internal/BaseHttpCommandExecutorService.java

```diff
@@ -80,7 +83,71 @@ protected BaseHttpCommandExecutorService(HttpUtils utils, ContentMetadataCodec c
       this.wire = checkNotNull(wire, "wire");
       this.idempotentMethods = ImmutableSet.copyOf(idempotentMethods.split(","));
    }
+   @Override
+   public ListenableFuture<HttpResponse> invokeAsync(final HttpCommand command){
+      SettableFuture<HttpResponse> finalFuture = SettableFuture.create();
+      this.invokeAsyncInternal(command, finalFuture);
+      return finalFuture;
+   }
+
+   private void invokeAsyncInternal(final HttpCommand command, final SettableFuture<HttpResponse> finalFuture) {
+      ListenableFuture<HttpResponse> future = null;
+      HttpRequest request = command.getCurrentRequest();
+      try {
+         for (HttpRequestFilter filter : request.getFilters()) {
+            request = filter.filter(request);
+         }
+         checkRequestHasContentLengthOrChunkedEncoding(request,
+                 "After filtering, the request has neither chunked encoding nor content length: " + request);
+         logger.debug("Sending request %s: %s", request.hashCode(), request.getRequestLine());
+         wirePayloadIfEnabled(wire, request);
+         utils.logRequest(headerLog, request, ">>");
+         final Q nativeRequest = convert(request);
+         future = invokeAsync(nativeRequest);
+         final int requestHashCode = request.hashCode();
```

Comment:
It is **very** unlikely that race conditions appear here, but better compute the hashcode before getting the future. Otherwise the callback could be executed while this is not yet assigned.